helpers/snapshot_tests_status.go

```diff
@@ -0,0 +1,191 @@
+			Status:       status,
+			Details:      details,
+			Timestamp:    timestamp,
+		}
+		sts.Statuses[scenarioName] = &newDetail
+		sts.Dirty = true
+	}
+
+}
+
+func (sts *SnaphostTestsStatus) InitStatuses(scenarios *[]v1beta1.IntegrationTestScenario) {
+	var expectedScenarios map[string]struct{}
+	expectedScenarios = make(map[string]struct{}) // map as a set
+
+	// if given scenario doesn't exist, create it in pending state
+	for _, s := range *scenarios {
+		expectedScenarios[s.Name] = struct{}{}
+		_, ok := sts.Statuses[s.Name]
+		if !ok {
+			// init test statuses only if they doesn't exist
+			sts.UpdateTestStatusIfChanged(s.Name, TestStatusPending, "Pending")
+		}
+	}
+
+	// remove old scenarios which are not defined anymore
+	for _, detail := range sts.Statuses {
+		_, ok := expectedScenarios[detail.ScenarioName]
+		if !ok {
+			sts.DeleteStatus(detail.ScenarioName)
+		}
+	}
+}
+
+func (sts *SnaphostTestsStatus) DeleteStatus(scenarioName string) {
+	_, ok := sts.Statuses[scenarioName]
+	if ok {
+		delete(sts.Statuses, scenarioName)
+		sts.Dirty = true
+	}
+}
+
+func (sts *SnaphostTestsStatus) MarshalJSON() ([]byte, error) {
+	result := make([]*TestStatusDetail, 0, len(sts.Statuses))
+	for _, v := range sts.Statuses {
+		result = append(result, v)
+	}
+	return json.Marshal(result)
+}
+
+func (sts *SnaphostTestsStatus) UnmarshalJSON(b []byte) error {
+	var inputData []*TestStatusDetail
+
+	err := json.Unmarshal(b, &inputData)
+	if err != nil {
+		return fmt.Errorf("Failed to unmarshal json data: %q", err)
+	}
+
+	for _, v := range inputData {
+		sts.Statuses[v.ScenarioName] = v
+	}
+
+	return nil
+}
+
+func NewSnaphostTestsStatus() *SnaphostTestsStatus {
+	sts := SnaphostTestsStatus{
+		Statuses: make(map[string]*TestStatusDetail, 1),
+		Dirty:    false,
+	}
+	return &sts
+}
+
+func NewSnapshotTestsStatusFromSnapshot(s applicationapiv1alpha1.Snapshot) (*SnaphostTestsStatus, error) {
+	annotations := map[string]string{}
+	if s.ObjectMeta.GetAnnotations() != nil {
+		annotations = s.ObjectMeta.GetAnnotations()
+	}
+	sts := NewSnaphostTestsStatus()
+
+	statusAnnotation, ok := annotations[SnapshotTestsStatusAnnotation]
+	if ok {
+		err := json.Unmarshal([]byte(statusAnnotation), sts)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to load tests statuses from the scenario annotation: %q", err)
+		}
+	}
+
+	return sts, nil
+}
+
+func WriteTestsStatusIntoSnapshot(s *applicationapiv1alpha1.Snapshot, sts *SnaphostTestsStatus, c client.Client, ctx context.Context) error {
+	if !sts.Dirty {
+		// No updates were done, we don't need to update snapshot
+		return nil
+	}
+	patch := client.MergeFrom(s.DeepCopy())
+
+	value, err := json.Marshal(sts)
+	if err != nil {
+		return fmt.Errorf("Failed to marshal test results into JSON: %q", err)
+	}
+	AddAnnotation(&s.ObjectMeta, SnapshotTestsStatusAnnotation, string(value))
+	err = c.Status().Patch(ctx, s, patch)
+	if err != nil {
+		return fmt.Errorf("Failed to update annotations: %q", err)
+	}
```

Comment:
I'm doing something wrong. This pass but snapshot object annotations are not updated